File: 2265-count-nodes-equal-to-average-of-subtree/2265-count-nodes-equal-to-average-of-subtree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def averageOfSubtree(self, root: Optional[TreeNode]) -> int:
        
        def dfs(root: Optional[TreeNode]) -> List[int]:
            if not root:
                return [0, 0, 0]
            res1, sum1, n1 = dfs(root.left)
            res2, sum2, n2 = dfs(root.right)
            sum3 = sum1 + sum2 + root.val
            n3 = n1 + n2 + 1
            avg = sum3 // n3
            res3 = res1 + res2 + (1 if avg == root.val else 0)
            return [res3, sum3, n3]
        
        return dfs(root)[0]

File: 2265-count-nodes-equal-to-average-of-subtree/2265-count-nodes-equal-to-average-of-subtree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var averageOfSubtree = function(root) {
    return dfs(root)[0]
    
    function dfs(root) {
        if (!root) return [0, 0, 0]
        const [cnt1, sum1, n1] = dfs(root.left)
        const [cnt2, sum2, n2] = dfs(root.right)
        let newCnt = cnt1 + cnt2
        const sum = sum1 + sum2 + root.val
        const n = n1 + n2 + 1
        const avg = Math.floor(sum / n)
        if (avg === root.val) newCnt++
        return [newCnt, sum, n]
    }
};